server/repository/repository.go

```diff
@@ -112,7 +115,27 @@ func (s *Server) Create(ctx context.Context, q *RepoCreateRequest) (*appsv1.Repo
 	if !s.enf.EnforceClaims(ctx.Value("claims"), "repositories", "create", fmt.Sprintf("*/%s", q.Repo.Repo)) {
 		return nil, grpc.ErrPermissionDenied
 	}
-	repo, err := s.db.CreateRepository(ctx, q.Repo)
+	r := q.Repo
+	repo, err := s.db.CreateRepository(ctx, r)
+	if status.Convert(err).Code() == codes.AlreadyExists {
+		// act idempotent if existing spec matches new spec
+		existing, getErr := s.db.GetRepository(ctx, r.Repo)
+		if getErr != nil {
+			return nil, status.Errorf(codes.Internal, "unable to check existing repository details: %v", err)
+		}
+		if q.Upsert {
+			return s.Update(ctx, &RepoUpdateRequest{Repo: r})
```

Comment:
Instead of doing the Update here i you see upsert flag specified, I think the logic should be reversed.

If you hit AlreadyExists, then perform a deep equal. If they are the same, then it is a noop (regardless if upsert was specified).

If deepequal fails, check upsert flag and decide to Update or fail according to the flag.

This way prevents changes to the secret, which triggers unnecessary watch events to fire as a result of the update.